compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertLoopsForVectorization.cpp

```diff
@@ -0,0 +1,172 @@
+    // operand 1 : (i, j, k) -> (k, j)
+    // result    : (i, j, k) -> (i, j)
+
+    auto indexingMaps = genericOp.getIndexingMaps();
+    assert(indexingMaps.size() == 3 && "expected 3 indexing maps here");
+
+    AffineMap map0 = cast<AffineMapAttr>(indexingMaps[0]).getValue();
+    AffineMap map1 = cast<AffineMapAttr>(indexingMaps[1]).getValue();
+    AffineMap map2 = cast<AffineMapAttr>(indexingMaps[2]).getValue();
+
+    auto nResults0 = map0.getNumResults();
+    auto nResults1 = map1.getNumResults();
+    auto nResults2 = map2.getNumResults();
+
+    if (map0.getResult(nResults0 - 2) != map2.getResult(nResults2 - 2))
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "unable to locate 'M' in affine maps [(M,K) x (K,N) -> (M,N)]");
+    if (map1.getResult(nResults1 - 1) != map2.getResult(nResults2 - 1))
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "unable to locate 'N' in affine maps [(M,K) x (K,N) -> (M,N)]");
+    if (map0.getResult(nResults0 - 1) != map1.getResult(nResults1 - 2))
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          "unable to locate 'K' in affine maps [(M,K) x (K,N) -> (M,N)]");
+
+    // Tile all dimensions except the 3 inner-most dimensions. Tile size '1'
+    // denotes tiling with smallest possible tile (size 1) and tile size '0'
+    // denotes tiling with the largest possible tile (size equal to the
+    // dimension size). The tiles sizes are thus [1,1,...1,0,0,0].
+    SmallVector<int64_t> tileSizes(numIterators, 1);
+    tileSizes[numIterators - 3] = 0;
+    tileSizes[numIterators - 2] = 0;
+    tileSizes[numIterators - 1] = 0;
+    auto outerOpts = linalg::LinalgTilingOptions().setTileSizes(tileSizes);
+    auto outerTiled = linalg::tileLinalgOp(rewriter, genericOp, outerOpts);
+    rewriter.replaceOp(genericOp, outerTiled.value().loops[0]->getResult(0));
+    return success();
+  }
+
+ private:
+  // Example: (1,1,4,1,2,3,4) has 5 dimensions from the first non-singleton.
+  static int getNumDimsFromFirstNonSingleton(Value v) {
+    auto shape = v.getType().cast<ShapedType>().getShape();
+    auto nDims = shape.size();
+    auto firstNonSingleton = [&]() -> int {
+      for (auto i : llvm::enumerate(shape)) {
+        if (i.value() > 1) {
+          return i.index();
+        }
+      }
+      return nDims;
+    }();
+    return nDims - firstNonSingleton;
+  };
+};
+
+class AMDAIEInsertLoopsForVectorizationPass
+    : public impl::AMDAIEInsertLoopsForVectorizationBase<
+          AMDAIEInsertLoopsForVectorizationPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, scf::SCFDialect>();
+  }
+  void runOnOperation() override;
+};
+
+void AMDAIEInsertLoopsForVectorizationPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto operation = getOperation();
+  IRRewriter rewriter(context);
+  RewritePatternSet patterns(operation.getContext());
```

Comment:
For these it is better to not use rewrite patterns. They are just one-off transformations. Rewriter is useful for fixed point, which this isnt. Just walk the program and find the operation you want to tile, and tile it once. That is much easier.

Reply:
Changed to getNumInnerDims